File: frontend/app/utils/role/permissions.ts
import type { AppRolePermissionRow } from '~/types/docetra/entities'

/** Canonical action codes accepted by the future authorization API. */
export const ROLE_PERMISSION_ACTIONS = [
  'view',
  'create',
  'edit',
  'delete',
  'assign',
  'share',
  'export',
  'comment',
  'transition',
  'configure',
] as const

export type RolePermissionAction = (typeof ROLE_PERMISSION_ACTIONS)[number]

export interface RoleDocumentTypeDefinition {
  value: string
  labelKey: string
  permissionPrefix: string
  actions: readonly RolePermissionAction[]
}

/** Matrix rows map to the same namespace used by page and API authorization. */
const WORKFLOW_ACTIONS: readonly RolePermissionAction[] = ['view', 'create', 'edit', 'delete', 'assign', 'share', 'export', 'comment', 'transition']
const MASTER_DATA_ACTIONS: readonly RolePermissionAction[] = ['view', 'create', 'edit', 'delete', 'export', 'comment']
const AUDIT_ACTIONS: readonly RolePermissionAction[] = ['view', 'export']
const CONFIG_ACTIONS: readonly RolePermissionAction[] = ['view', 'create', 'edit', 'delete', 'comment', 'configure']
export const ROLE_DOCUMENT_TYPES: readonly RoleDocumentTypeDefinition[] = [
  { value: 'incoming_document', labelKey: 'docetra.pages.incomingDocument', permissionPrefix: 'records.incoming_documents', actions: WORKFLOW_ACTIONS },
  { value: 'outgoing_document', labelKey: 'docetra.pages.outgoingDocument', permissionPrefix: 'records.outgoing_documents', actions: WORKFLOW_ACTIONS },
  { value: 'document', labelKey: 'docetra.pages.document', permissionPrefix: 'records.documents', actions: WORKFLOW_ACTIONS },
  { value: 'master_list_request', labelKey: 'docetra.pages.masterListRequest', permissionPrefix: 'records.master_list_requests', actions: WORKFLOW_ACTIONS },
  { value: 'meeting_topic', labelKey: 'docetra.pages.meetingTopic', permissionPrefix: 'meetings.topics', actions: WORKFLOW_ACTIONS },
  { value: 'meeting_history', labelKey: 'docetra.pages.meetingHistory', permissionPrefix: 'meetings.history', actions: WORKFLOW_ACTIONS },
  { value: 'department', labelKey: 'docetra.pages.department', permissionPrefix: 'organizations.departments', actions: MASTER_DATA_ACTIONS },
  { value: 'company', labelKey: 'docetra.pages.company', permissionPrefix: 'organizations.companies', actions: MASTER_DATA_ACTIONS },
  { value: 'company_purpose', labelKey: 'docetra.pages.companyPurpose', permissionPrefix: 'organizations.company_purposes', actions: MASTER_DATA_ACTIONS },
  { value: 'company_sector', labelKey: 'docetra.pages.companySector', permissionPrefix: 'organizations.company_sectors', actions: MASTER_DATA_ACTIONS },
  { value: 'officer', labelKey: 'docetra.pages.officer', permissionPrefix: 'organizations.officers', actions: MASTER_DATA_ACTIONS },
  { value: 'user', labelKey: 'docetra.pages.user', permissionPrefix: 'users.users', actions: ['view', 'create', 'edit', 'delete', 'configure'] },
  { value: 'role', labelKey: 'docetra.pages.role', permissionPrefix: 'users.roles', actions: ['view', 'create', 'edit', 'delete', 'configure'] },
  { value: 'record_type', labelKey: 'docetra.pages.recordType', permissionPrefix: 'configuration.record_types', actions: CONFIG_ACTIONS },
  { value: 'record_attribute', labelKey: 'docetra.pages.recordAttribute', permissionPrefix: 'configuration.record_attributes', actions: CONFIG_ACTIONS },
  { value: 'record_log', labelKey: 'docetra.pages.recordLog', permissionPrefix: 'records.logs', actions: AUDIT_ACTIONS },
  { value: 'file_upload', labelKey: 'docetra.pages.fileUpload', permissionPrefix: 'portal.file_upload', actions: ['view', 'create', 'delete', 'share', 'export'] },
  { value: 'google_drive_sync', labelKey: 'docetra.pages.googleDriveSync', permissionPrefix: 'portal.google_drive_sync', actions: CONFIG_ACTIONS },
  { value: 'portal_log', labelKey: 'docetra.pages.portalLog', permissionPrefix: 'portal.logs', actions: AUDIT_ACTIONS },
  { value: 'system_log', labelKey: 'docetra.pages.systemLog', permissionPrefix: 'system.logs', actions: AUDIT_ACTIONS },
  { value: 'app_config', labelKey: 'docetra.pages.appConfig', permissionPrefix: 'settings.app_config', actions: ['view', 'edit', 'configure'] },
  { value: 'app_info', labelKey: 'docetra.pages.appInfo', permissionPrefix: 'settings.app_info', actions: ['view', 'edit', 'configure'] },
  { value: 'storage', labelKey: 'docetra.pages.storage', permissionPrefix: 'settings.storage', actions: ['view', 'edit', 'configure'] },
] as const

const ACTION_SET = new Set<string>(ROLE_PERMISSION_ACTIONS)
const LEGACY_ACTION_MAP: Record<string, RolePermissionAction | undefined> = {
  select: 'view',
  read: 'view',
  write: 'edit',
  email: 'comment',
  report: 'view',
  import: 'create',
  mask: 'view',
}

export function normalizePermissionActions(actions: readonly string[] | null | undefined): RolePermissionAction[] {
  const normalized = new Set<RolePermissionAction>()
  for (const raw of actions || []) {
    const action = ACTION_SET.has(raw)
      ? raw as RolePermissionAction
      : LEGACY_ACTION_MAP[raw]
    if (action) normalized.add(action)
  }
  if ([...normalized].some(action => action !== 'view')) normalized.add('view')
  return ROLE_PERMISSION_ACTIONS.filter(action => normalized.has(action))
}

export function createEmptyPermissionRow(documentType: string): AppRolePermissionRow {
  return {
    id: `perm_${documentType}`,
    documentType,
    onlyIfCreator: false,
    level: 0,
    actions: [],
  }
}

/** Merge API rows with the current matrix catalog and discard unknown rows/actions. */
export function normalizePermissionRows(
  rows: readonly AppRolePermissionRow[] | null | undefined,
  includeEmpty = true,
): AppRolePermissionRow[] {
  const byType = new Map((rows || []).map(row => [row.documentType, row]))
  const normalized = ROLE_DOCUMENT_TYPES.map((definition) => {
    const existing = byType.get(definition.value)
    const actions = normalizePermissionActions(existing?.actions)
      .filter(action => definition.actions.includes(action))
    return {
      id: existing?.id || `perm_${definition.value}`,
      documentType: definition.value,
      onlyIfCreator: actions.length ? Boolean(existing?.onlyIfCreator) : false,
      level: Math.min(9, Math.max(0, Number(existing?.level || 0))),
      actions,
    }
  })
  return includeEmpty ? normalized : normalized.filter(row => row.actions.length > 0)
}

/** Enforce action dependencies consistently for checkbox and API payload flows. */
export function setPermissionAction(
  row: AppRolePermissionRow,
  action: string,
  enabled: boolean,
): AppRolePermissionRow {
  const normalizedAction = ACTION_SET.has(action)
    ? action as RolePermissionAction
    : LEGACY_ACTION_MAP[action]
  if (!normalizedAction) return row
  const actions = new Set(normalizePermissionActions(row.actions))
  if (enabled) {
    actions.add(normalizedAction)
    actions.add('view')
  }
  else if (normalizedAction === 'view') {
    actions.clear()
  }
  else {
    actions.delete(normalizedAction)
  }
  const ordered = ROLE_PERMISSION_ACTIONS.filter(item => actions.has(item))
  return {
    ...row,
    actions: ordered,
    onlyIfCreator: ordered.length ? Boolean(row.onlyIfCreator) : false,
  }
}

/** Expanded capabilities sent with structured rows for fast authorization checks. */
export function permissionRowsToFlatKeys(rows: AppRolePermissionRow[]): string[] {
  const definitions = new Map(ROLE_DOCUMENT_TYPES.map(item => [item.value, item]))
  const keys = new Set<string>()
  for (const row of normalizePermissionRows(rows, false)) {
    const prefix = definitions.get(row.documentType)?.permissionPrefix
    if (!prefix) continue
    for (const action of row.actions) keys.add(`${prefix}.${action}`)
  }
  return [...keys].sort()
}

File: backend/app/core/permissions.py
WORKFLOW = ["view", "create", "edit", "archive", "restore", "delete", "purge", "assign", "share", "export", "comment", "transition"]
MASTER = ["view", "create", "edit", "delete", "export", "comment"]
AUDIT = ["view", "export"]
CONFIG = ["view", "create", "edit", "delete", "export", "comment", "configure"]
SETTINGS = ["view", "edit", "configure"]
USERS = ["view", "create", "edit", "archive", "restore", "delete", "purge", "comment", "configure"]

# Non-record permission rows (orgs, users, settings, portal, …).
STATIC_DOCUMENT_TYPES: list[tuple[str, str, list[str]]] = [
    ("dashboard", "dashboard", ["view"]),
    ("archive", "archive", ["view"]),
    ("department", "organizations.departments", MASTER),
    ("company", "organizations.companies", MASTER),
    ("purpose", "organizations.purposes", MASTER),
    ("sector", "organizations.sectors", MASTER),
    ("officer", "organizations.officers", MASTER),
    ("user", "users.users", USERS),
    ("role", "users.roles", USERS),
    ("record_type", "configuration.record_types", CONFIG),
    ("record_attribute", "configuration.record_attributes", CONFIG),
    ("record_log", "records.logs", AUDIT),
    ("file_upload", "portal.file_upload", ["view", "create", "delete", "share", "export"]),
    ("google_drive_sync", "portal.google_drive_sync", CONFIG),
    ("portal_log", "portal.logs", AUDIT),
    ("system_log", "system.logs", AUDIT),
    ("app_config", "settings.app_config", SETTINGS),
    ("app_info", "settings.app_info", SETTINGS),
    ("storage", "settings.storage", SETTINGS),
]


def _record_document_types() -> list[tuple[str, str, list[str]]]:
    """Generate record-type permission rows from built-in type codes (no hand-edited DOCUMENT_TYPES)."""
    from app.modules.record.domain.map import TYPE_UI_DEFAULTS, permission_prefix_for_type_code

    return [(code, permission_prefix_for_type_code(code), WORKFLOW) for code in TYPE_UI_DEFAULTS]


def _all_document_types() -> list[tuple[str, str, list[str]]]:
    # Record types first so catalog order stays familiar.
    return [*_record_document_types(), *STATIC_DOCUMENT_TYPES]


# Lazy-compatible aliases used across the codebase.
DOCUMENT_TYPES = _all_document_types()
PREFIXES = {prefix: actions for _, prefix, actions in DOCUMENT_TYPES}
TYPE_TO_PREFIX = {doc_type: prefix for doc_type, prefix, _ in DOCUMENT_TYPES}
ALL_PERMISSIONS = [f"{prefix}.{action}" for prefix, actions in PREFIXES.items() for action in actions]


def refresh_permission_tables() -> None:
    """Rebuild PREFIXES/TYPE_TO_PREFIX/ALL_PERMISSIONS (e.g. after new type seeds)."""
    global DOCUMENT_TYPES, PREFIXES, TYPE_TO_PREFIX, ALL_PERMISSIONS
    DOCUMENT_TYPES = _all_document_types()
    PREFIXES = {prefix: actions for _, prefix, actions in DOCUMENT_TYPES}
    TYPE_TO_PREFIX = {doc_type: prefix for doc_type, prefix, _ in DOCUMENT_TYPES}
    ALL_PERMISSIONS = [f"{prefix}.{action}" for prefix, actions in PREFIXES.items() for action in actions]


def normalize_permission_payload(payload: dict) -> dict:
    refresh_permission_tables()
    rows = payload.get("permissionRows") or []
    normalized = []
    flat = set()
    for raw in rows:
        document_type = str(raw.get("documentType", ""))
        prefix = TYPE_TO_PREFIX.get(document_type)
        if not prefix:
            # Allow new DB type codes: records.{type_code}.*
            if document_type and document_type not in {t for t, _, _ in STATIC_DOCUMENT_TYPES}:
                prefix = f"records.{document_type}"
                allowed = set(WORKFLOW)
            else:
                raise ValueError(f"Unknown permission document type: {document_type}")
        else:
            allowed = set(PREFIXES.get(prefix) or WORKFLOW)
        actions = []
        for action in raw.get("actions") or []:
            if action not in allowed:
                raise ValueError(f"Action {action} is not allowed for {document_type}")
            actions.append(action)
        if actions and "view" not in actions:
            actions.insert(0, "view")
        actions = list(dict.fromkeys(actions))
        level = int(raw.get("level", 0))
        if not 0 <= level <= 9:
            raise ValueError("Permission level must be between 0 and 9")
        normalized.append({
            "id": raw.get("id") or f"perm_{document_type}",
            "documentType": document_type,
            "actions": actions,
            "onlyIfCreator": bool(raw.get("onlyIfCreator")) if actions and "purge" not in actions else False,
            "level": level,
        })
        flat.update(f"{prefix}.{action}" for action in actions)
    computed = sorted(flat)
    supplied = payload.get("permissions")
    if supplied is not None and sorted(set(supplied)) != computed:
        raise ValueError("permission_payload_mismatch")
    return {**payload, "permissionRows": normalized, "permissions": computed, "permissionCount": len(computed), "permissionSchemaVersion": 1}


def catalog_rows() -> list[dict]:
    refresh_permission_tables()
    return [
        {
            "id": f"perm_{doc_type}",
            "documentType": doc_type,
            "actions": list(actions),
            "onlyIfCreator": False,
            "level": 0,
        }
        for doc_type, _prefix, actions in DOCUMENT_TYPES
    ]


def expand_permission_rows(rows: list | None) -> list[str]:
    refresh_permission_tables()
    keys: list[str] = []
    for row in rows or []:
        if not isinstance(row, dict):
            continue
        document_type = str(row.get("documentType") or "")
        prefix = TYPE_TO_PREFIX.get(document_type) or (f"records.{document_type}" if document_type else "")
        if not prefix:
            continue
        for action in row.get("actions") or []:
            keys.append(f"{prefix}.{action}")
    return keys
